src/main/java/com/uid2/admin/vertx/service/SharingService.java

```diff
@@ -0,0 +1,187 @@
+        jo.put("whitelist", listedSites);
+        jo.put("whitelist_hash", computeWhitelistHash(acl.getAccessList()));
+
+        rc.response()
+                .putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                .end(jo.encode());
+    }
+
+    private void handleKeyAclListAll(RoutingContext rc) {
+        try {
+            JsonArray ja = new JsonArray();
+            Map<Integer, EncryptionKeyAcl> collection = this.keyAclProvider.getSnapshot().getAllAcls();
+            for (Map.Entry<Integer, EncryptionKeyAcl> acl : collection.entrySet()) {
+                JsonArray listedSites = new JsonArray();
+                acl.getValue().getAccessList().stream().sorted().forEach((listedSiteId) -> listedSites.add(listedSiteId));
+
+                JsonObject jo = new JsonObject();
+                jo.put("site_id", acl.getKey());
+                jo.put("whitelist", listedSites);
+                jo.put("whitelist_hash", computeWhitelistHash(acl.getValue().getAccessList()));
+                ja.add(jo);
+            }
+
+            rc.response()
+                    .putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                    .end(ja.encode());
+        } catch (Exception e) {
+            rc.fail(500, e);
+        }
+    }
+
+    private void handleKeyAclSet(RoutingContext rc) {
+        int site_id;
+        try {
+            site_id = Integer.parseInt(rc.pathParam("siteId"));
+        } catch (Exception e) {
+            LOGGER.warn("Failed to parse a site id from list request", e);
+            rc.fail(400, e);
+            return;
+        }
+
+        try {
+            keyAclProvider.loadContent();
+        } catch (Exception e) {
+            LOGGER.error("Failed to load key acls");
+            rc.fail(500);
+        }
+
+
+        final Map<Integer, EncryptionKeyAcl> collection = this.keyAclProvider.getSnapshot().getAllAcls();
+        EncryptionKeyAcl acl;
+        try {
+            acl = collection.get(site_id);
+        } catch (Exception e) {
+            LOGGER.warn("Failed to find acl for site id: " + site_id, e);
+            rc.fail(404, e);
+            return;
+        }
+
+        final JsonObject body = new JsonObject(rc.body().asString());
+
+        final JsonArray whitelist = body.getJsonArray("whitelist");
+        final int whitelist_hash = body.getInteger("whitelist_hash");
+
+
+        Set<Integer> old_list;
+        if (acl != null) {
+            old_list = acl.getAccessList();
+        } else {
+            old_list = new HashSet<>();
+        }
+
+        if(acl != null && whitelist_hash != computeWhitelistHash(old_list)) {
+            rc.fail(409);
+            return;
+        }
+
+        final EncryptionKeyAcl newAcl = new EncryptionKeyAcl(true, whitelist.stream()
+                .map(s -> (Integer) s)
+                .collect(Collectors.toSet()));
+
+        collection.put(site_id, newAcl);
+        try {
+            storeWriter.upload(collection, null);
```

Comment:
Do we need some locking from the moment we read until when we write back? Multiple requests could come in at around the same time. I see there's a writeLock injected but it doesn't seem to be used.